### ISSUE TITLE
dnf5 performance vs dnf: should it be better?

### ISSUE THREAD

**Issue by jashank** [rpm-software-management/dnf5#1251]
Hello!

I've mostly migrated from `dnf` 4 (currently 4.19.0) to `dnf5` (currently 5.11.12) on my rawhide system. I'm sure I've seen references to the design being intentionally faster — and I've been certainly enjoying places where `dnf5` performs way better than `dnf` — e.g., responding to `C-c` pretty rapidly.  But I've found `dnf5` _feels_ slower.

<details>
<summary><code>/etc/dnf/dnf.conf</code></summary>

```
[main]
gpgcheck = True
installonly_limit = 3
clean_requirements_on_remove = True
best = False
skip_if_unavailable = True
keepcache = True

countme = True

fastestmirror = True
max_parallel_downloads = 10

# Try to never, ever remove package management logs.
log_rotate = 64
log_compress = True
```

</details>

<details>
<summary><code>rpm -qa | grep 'dnf|hawkey|libsolv'</code></summary>

```
libsolv-0.7.28-1.fc40.x86_64
libdnf5-5.1.12-1.fc40.x86_64
libdnf5-cli-5.1.12-1.fc40.x86_64
dnf-data-4.19.0-1.fc40.noarch
libdnf-0.73.0-1.fc40.x86_64
python3-libdnf-0.73.0-1.fc40.x86_64
python3-hawkey-0.73.0-1.fc40.x86_64
dnf5-5.1.12-1.fc40.x86_64
dnf5daemon-server-5.1.12-1.fc40.x86_64
python3-dnf-4.19.0-1.fc40.noarch
python3-dnf-plugins-core-4.5.0-1.fc40.noarch
dnf-plugins-core-4.5.0-1.fc40.noarch
dnf-4.19.0-1.fc40.noarch
dnf-utils-4.5.0-1.fc40.noarch
python3-dnf-plugin-versionlock-4.5.0-1.fc40.noarch
python3-dnf-plugin-post-transaction-actions-4.5.0-1.fc40.noarch
python3-dnf-plugin-modulesync-4.5.0-1.fc40.noarch
dnf5daemon-client-5.1.12-1.fc40.x86_64
dnf5-plugins-5.1.12-1.fc40.x86_64
libsolv-demo-0.7.28-1.fc40.x86_64
ruby-libdnf5-cli-5.1.12-1.fc40.x86_64
perl-libdnf5-cli-5.1.12-1.fc40.x86_64
ruby-libdnf5-5.1.12-1.fc40.x86_64
perl-libdnf5-5.1.12-1.fc40.x86_64
libsolv-tools-0.7.28-1.fc40.x86_64
```

</details>

And, empirically[^1]:

```shell-session
% hyperfine --shell=none -- \
    'dnf -C info hplip' \
    'dnf5 -C info hplip'
Benchmark 1: dnf -C info hplip
  Time (mean ± σ):      9.010 s ±  0.166 s    [User: 3.554 s, System: 5.178 s]
  Range (min … max):    8.853 s …  9.341 s    10 runs
 
Benchmark 2: dnf5 -C info hplip
  Time (mean ± σ):     10.335 s ±  0.138 s    [User: 5.233 s, System: 5.151 s]
  Range (min … max):   10.125 s … 10.536 s    10 runs
 
Summary
  dnf -C info hplip ran
    1.15 ± 0.03 times faster than dnf5 -C info hplip
```

I'm sure I've seen even worse performance, too :'(

This felt so frustrating that, when I hacked up a pile of Ruby that dismantled `dumpsolv -j` output into a SQLite3 database, and tried querying against that (here, `solvsql-info` parses all my extremely-nonoptimal Ruby too):

```shell-session
% hyperfine --shell=none -- \
    'dnf -C info hplip' \
    'dnf5 -C info hplip' \
    'solvsql-info hplip' \
    "sqlite3 /tmp/solv.db '.mode line' \"SELECT * FROM q_info WHERE Name = 'hplip'\"" 
Benchmark 1: dnf -C info hplip
  Time (mean ± σ):      8.706 s ±  0.161 s    [User: 3.378 s, System: 5.058 s]
  Range (min … max):    8.493 s …  8.944 s    10 runs
 
Benchmark 2: dnf5 -C info hplip
  Time (mean ± σ):     10.452 s ±  0.270 s    [User: 5.293 s, System: 5.187 s]
  Range (min … max):   10.193 s … 11.061 s    10 runs
 
Benchmark 3: solvsql-info hplip
  Time (mean ± σ):     358.2 ms ±  21.2 ms    [User: 252.0 ms, System: 95.6 ms]
  Range (min … max):   328.6 ms … 399.1 ms    10 runs
 
Benchmark 4: sqlite3 /tmp/solv.db '.mode line' "SELECT * FROM q_info WHERE Name = 'hplip'"
  Time (mean ± σ):       5.7 ms ±   0.6 ms    [User: 1.8 ms, System: 3.8 ms]
  Range (min … max):     4.4 ms …   7.9 ms    270 runs
 
Summary
  sqlite3 /tmp/solv.db '.mode line' "SELECT * FROM q_info WHERE Name = 'hplip'" ran
   62.30 ± 7.47 times faster than solvsql-info hplip
 1514.17 ± 160.14 times faster than dnf -C info hplip
 1817.97 ± 195.05 times faster than dnf5 -C info hplip
```

(I am not including the ~45s to load the libsolv databases, and I have some simple ideas for how to make that go even faster.)

I see similar characteristics for `dnf search` vs `dnf5 search` (and am even more startled by how slow things seem to be):

```shell-session
% hyperfine --shell=none -- \
    'dnf -C search printer' \
    'dnf5 -C search printer' \
    'solvsql-search printer' \
    "sqlite3 /tmp/solv.db \"SELECT FORMAT('%s.%s: %s', s.name, s.arch, s.summary) FROM solvables AS s WHERE s.name LIKE '%printer%' OR s.summary LIKE '%printer%'\""
Benchmark 1: dnf -C search printer
  Time (mean ± σ):      8.590 s ±  0.121 s    [User: 3.372 s, System: 4.982 s]
  Range (min … max):    8.405 s …  8.767 s    10 runs
 
Benchmark 2: dnf5 -C search printer
  Time (mean ± σ):      9.960 s ±  0.074 s    [User: 5.038 s, System: 5.003 s]
  Range (min … max):    9.879 s … 10.092 s    10 runs
 
Benchmark 3: solvsql-search printer
  Time (mean ± σ):     331.5 ms ±  17.3 ms    [User: 229.1 ms, System: 92.8 ms]
  Range (min … max):   309.5 ms … 362.1 ms    10 runs
 
Benchmark 4: sqlite3 /tmp/solv.db "SELECT FORMAT('%s.%s: %s', s.name, s.arch, s.summary) FROM solvables AS s WHERE s.name LIKE '%printer%' OR s.summary LIKE '%printer%'"
  Time (mean ± σ):      61.9 ms ±   2.7 ms    [User: 36.7 ms, System: 23.3 ms]
  Range (min … max):    59.5 ms …  72.0 ms    44 runs
 
Summary
  sqlite3 /tmp/solv.db "SELECT FORMAT('%s.%s: %s', s.name, s.arch, s.summary) FROM solvables AS s WHERE s.name LIKE '%printer%' OR s.summary LIKE '%printer%'" ran
    5.35 ± 0.36 times faster than solvsql-search printer
  138.70 ± 6.39 times faster than dnf -C search printer
  160.82 ± 7.15 times faster than dnf5 -C search printer
```

(And yes, I appreciate that the query filters here probably aren't quite right; but indicatively, that's what I'd expect to see. And that's without any sort of effort in tuning up the database — again, I have some simple ideas and partial implementations, using some well-chosen indices and the FTS5 extension.)

I've not started digging into `dnf5` performance yet — but I would not have expected to see `dnf5` perform quite as much slower...?  Is there some fundamental reason that it shouldn't be at least as fast as the tool it replaces?  And is there some reason that there is still such a huge disparity against, for example, simple SQLite searches over the same data?

[^1]: `dnf-4.19.0-1.fc40.noarch` vs `dnf5-5.1.12-1.fc40.x86_64` from Fedora rawhide modulo mainline kernel `6.6.10-200.fc39.x86_64`, on an AMD Ryzen 7 4750U throttled to 1.4 GHz all-core, on battery, under other loads — I'm not trying for rigour here, but rather indicative and comparative numbers.

**Comment by ppisar** [comment]
I only tried `-C info hplip` in Fedora 41 with developmental dnf4 and dnf5 and my results are quite different:

~~~~
# hyperfine --shell=none -- '/usr/bin/dnf-3 -C info hplip' '/usr/bin/dnf5 -C info hplip'
Benchmark 1: /usr/bin/dnf-3 -C info hplip
  Time (mean ± σ):     590.3 ms ±   7.5 ms    [User: 506.8 ms, System: 80.2 ms]
  Range (min … max):   574.9 ms … 598.0 ms    10 runs
 
Benchmark 2: /usr/bin/dnf5 -C info hplip
  Time (mean ± σ):     578.4 ms ±   2.9 ms    [User: 580.8 ms, System: 78.9 ms]
  Range (min … max):   575.1 ms … 584.6 ms    10 runs
 
Summary
  /usr/bin/dnf5 -C info hplip ran
    1.02 ± 0.01 times faster than /usr/bin/dnf-3 -C info hplip
~~~~

I observe 1.02 speedup for DNF5. You observe 1.15 speedup for DNF4.

(Though I don't think that querying details about a single package, i.e. extracting 805 bytes from dozens of compressed megabytes input is any representative workload.)

I don't say there are no places for an improvement. E.g. the libsolv database is not a text-oriented database as SQLite. libsolv is optimized for a dependency resolution. Not for retrieving texts. We can look where the bottlenecks are in DNF5, we can consider adding a new index for textual data, e.g. the SQLite. But this is not something top important right now. Especially when I don't observe your problem.

**Comment by jashank** [comment]
Thanks for the response!

Maybe I do need to dig further, given this seems to be a me problem.

Picking through the output of _strace(1)_ (`strace -tt -ff -o x $dnf -C search hplip`, so again, indicative and comparative), I see a few threads I would like to pull on further:

<details>
<summary>
dnf4 and dnf5 <em>read(2)</em>s repodata zchunk files sequentially by chunk
</summary>

unsurprising, given the zchunk format, but still startling to see, e.g.,

```
09:27:12.622787 read(4, "(\265/\375#0\321\321u\240\345\2\0\364\2<?xml=\"UTF-8\"?>\n<"..., 105) = 105
09:27:12.622940 read(4, "(\265/\375c0\321\321u{\27\315\34\0\343\226E\17\345k\207\200\321\303\307\f\322\210\0163d\303"..., 935) = 935
09:27:12.623084 read(4, "(\265/\375c0\321\321u\200\5\r\17\0cN)\33\24?\343)\343\7\343\2566\262\334j\2151"..., 495) = 495
09:27:12.623245 read(4, "(\265/\375c0\321\321uJ\5e\v\0\223\213!\222\332\254e\261\214,\356\267\360\275u\360\361r"..., 378) = 378
09:27:12.623406 read(4, "(\265/\375c0\321\321u+\20\325\16\0\303P*\224\212\244\214\245\252\206P\212R\20\245\230\361\372"..., 488) = 488
```

This seems like quite poor form: it takes ~5s to load rawhide's `primary.xml.zck`.

(Off-hand: I know the repodata used to include prerolled SQLite databases, and I wonder if, say, https://sqlite.org/sessionintro.html would make sense here.)

</details>

<details>
<summary>
dnf5 slow-<em>brk(2)</em>s whilst reading <code>nevras.toml</code>
</summary>

This one has me baffled:

```
09:27:10.535150 newfstatat(AT_FDCWD, "/usr/lib/sysimage/libdnf5/nevras.toml", {st_mode=S_IFREG|0644, st_size=707084, ...}, 0) = 0
09:27:10.535486 openat(AT_FDCWD, "/usr/lib/sysimage/libdnf5/nevras.toml", O_RDONLY) = 3
09:27:10.537185 lseek(3, 0, SEEK_CUR)   = 0
09:27:10.537571 lseek(3, 0, SEEK_END)   = 707084
09:27:10.537852 lseek(3, 0, SEEK_CUR)   = 707084
09:27:10.538084 lseek(3, 0, SEEK_SET)   = 0
09:27:10.538447 read(3, "version = \"1.0\"\n\n[nevras]\n\"4th-3"..., 707084) = 707084
09:27:11.984051 brk(0x563f7f624000)     = 0x563f7f624000
09:27:12.015178 brk(0x563f7f645000)     = 0x563f7f645000
09:27:12.045528 brk(0x563f7f666000)     = 0x563f7f666000
09:27:12.074173 brk(0x563f7f687000)     = 0x563f7f687000
```

I'm not sure why grabbing ~33 pages seems to take that long!  Every following sbrk happens in quick succession with ≤ 0.01s between calls, and I think I need to dig into why.  (I'm not under memory pressure, AFAIK...)

But I'm also somewhat startled here by TOML files for this purpose.  I suspect I'd need to go dig through dnf5 code to work out what's going on, and whether it needs to be this way.  (Not for any lack of love for TOML ... but loading ~1 MiB of TOML seems perhaps suboptimal?)


</details>

<details>
<summary>
dnf5 logs via <em>open(2)</em>/<em>flock(2)</em>/<em>lseek(2)</em>/<em>write(2)</em>/<em>close(2)</em>
</summary>

Not especially noteworthy for performance, but a strange observation nonetheless:

```
09:27:12.275593 openat(AT_FDCWD, "/home/jashank/.local/state/dnf5.log", O_WRONLY|O_CREAT|O_APPEND|O_CLOEXEC, 0644) = 4
09:27:12.275732 flock(4, LOCK_EX)       = 0
09:27:12.275853 lseek(4, 0, SEEK_END)   = 84407
09:27:12.275954 write(4, "2024-02-19T22:27:09+0000 [285951"..., 129) = 129
09:27:12.276108 close(4)                = 0
```

```shell-session
% grep '/home/jashank/.local/state/dnf5.log' * | wc
    461    3227   59483
```

</details>

<details>
<summary>
reading <code>rawhide.solv</code> is sort-of slow?
</summary>

reading the libsolv database seems to go in two phases: one that does lots of sequential reads from the top of the file for ~0.8s, then one that does lots of big skips and read for another ~0.2s.  (I assume it's walking some sort of chunk-list, but I haven't looked at libsolv's file format.)

is there any reason this is done by reads and not by, say, mmap'ing the files in?

</details>

<details>
<summary>
reading the rpmdb is <em>very</em> slow
</summary>

_strace(2)_ suggested lots of reads that ended up being the rpmdb.  Okay: why is _that_ slow?

```shell-session
% hyperfine --shell=none -- 'rpm -qa'
Benchmark 1: rpm -qa
  Time (mean ± σ):     10.634 s ±  0.152 s    [User: 5.920 s, System: 4.430 s]
  Range (min … max):   10.489 s … 10.915 s    10 runs
```

On a hunch:

```shell-session
% l /usr/lib/sysimage/rpm/
total 618M
-rw-r--r--. 1 root root 1112858624 Feb 19 23:03 rpmdb.sqlite
-rw-r--r--. 1 root root      32768 Feb 19 23:03 rpmdb.sqlite-shm
-rw-r--r--. 1 root root          0 Feb 19 23:03 rpmdb.sqlite-wal
```

I've seen a few remarks about file size after the Berkeley DB → SQLite3 migration, which strike me as surprising. (I'm also not entirely sure the set of indices on the SQLite3-shaped rpmdb are conducive to text searches.)

```shell-session
% sudo sqlite3 /usr/lib/sysimage/rpm/rpmdb.sqlite
SQLite version 3.45.1 2024-01-30 16:01:20
Enter ".help" for usage hints.

sqlite> .stats
Memory Used:                         65192 (max 65280) bytes
Number of Outstanding Allocations:   157 (max 158)
Number of Pcache Overflow Bytes:     4104 (max 4104) bytes
Largest Allocation:                  48000 bytes
Largest Pcache Allocation:           4104 bytes
Lookaside Slots Used:                41 (max 41)
Successful lookaside attempts:       41
Lookaside failures due to size:      0
Lookaside failures due to OOM:       0
Pager Heap Usage:                    4984 bytes
Page cache hits:                     0
Page cache misses:                   0
Page cache writes:                   0
Page cache spills:                   0
Schema Heap Usage:                   0 bytes
Statement Heap/Lookaside Usage:      0 bytes
Bytes received by read():            30538
Bytes sent to write():               4776
Read() system calls:                 36
Write() system calls:                98
Bytes read from storage:             51032
Bytes written to storage:            0
Cancelled write bytes:               0

sqlite> .dbinfo
database page size:  4096
write format:        2
read format:         2
reserved bytes:      0
file change counter: 315
database page count: 271694
freelist page count: 109332
schema cookie:       49
schema format:       4
default cache size:  0
autovacuum top root: 0
incremental vacuum:  0
text encoding:       1 (utf8)
user version:        0
application id:      0
software version:    3045001
number of tables:    21
number of indexes:   29
number of triggers:  0
number of views:     0
schema size:         4491
data version         2

sqlite> VACUUM;

sqlite> PRAGMA OPTIMIZE;

sqlite> .dbinfo
database page size:  4096
write format:        2
read format:         2
reserved bytes:      0
file change counter: 316
database page count: 150068
freelist page count: 0
schema cookie:       50
schema format:       4
default cache size:  0
autovacuum top root: 0
incremental vacuum:  0
text encoding:       1 (utf8)
user version:        0
application id:      0
software version:    3045001
number of tables:    21
number of indexes:   29
number of triggers:  0
number of views:     0
schema size:         4491
data version         3

sqlite> .stats
Memory Used:                         2287960 (max 4522072) bytes
Number of Outstanding Allocations:   881 (max 1594)
Number of Pcache Overflow Bytes:     2029920 (max 4076800) bytes
Largest Allocation:                  309520 bytes
Largest Pcache Allocation:           4368 bytes
Lookaside Slots Used:                46 (max 123)
Successful lookaside attempts:       5537
Lookaside failures due to size:      74
Lookaside failures due to OOM:       651
Pager Heap Usage:                    2111888 bytes
Page cache hits:                     72
Page cache misses:                   312426
Page cache writes:                   150068
Page cache spills:                   149634
Schema Heap Usage:                   15208 bytes
Statement Heap/Lookaside Usage:      0 bytes
Bytes received by read():            2507536480
Bytes sent to write():               1848066693
Read() system calls:                 612271
Write() system calls:                600838
Bytes read from storage:             2507536300
Bytes written to storage:            1849316940
Cancelled write bytes:               0

sqlite> .q

% sudo rpm -qa >/dev/null
% l /usr/lib/sysimage/rpm/
total 330M
-rw-r--r--. 1 root root 614678528 Feb 20 09:42 rpmdb.sqlite
-rw-r--r--. 1 root root     32768 Feb 20 09:50 rpmdb.sqlite-shm
-rw-r--r--. 1 root root         0 Feb 20 09:50 rpmdb.sqlite-wal
```

(Invoking `rpm -qa` here was a fast and easy way to put the `-shm` and `-wal` files back as root.  Without that, when RPM preens the schema, SQLite3 bails out claiming "attempt to write a readonly database".)

Halving the size of the `rpmdb` _helps_, but not much:

```shell-session
% hyperfine --shell=none -- 'rpm -qa'
Benchmark 1: rpm -qa
  Time (mean ± σ):      7.777 s ±  0.093 s    [User: 5.835 s, System: 1.772 s]
  Range (min … max):    7.638 s …  7.978 s    10 runs
```

It also doesn't move the needle on the dnf4 vs dnf5 numbers much:

```
^ hyperfine --shell=none -- 'dnf -C search printer' 'dnf5 -C search printer'
Benchmark 1: dnf -C search printer
  Time (mean ± σ):      5.874 s ±  0.096 s    [User: 3.295 s, System: 2.423 s]
  Range (min … max):    5.751 s …  5.999 s    10 runs

Benchmark 2: dnf5 -C search printer
  Time (mean ± σ):      7.185 s ±  0.047 s    [User: 4.974 s, System: 2.403 s]
  Range (min … max):    7.125 s …  7.261 s    10 runs

Summary
  dnf -C search printer ran
    1.22 ± 0.02 times faster than dnf5 -C search printer
```

I might dig further into this one, given it seems there's something suspect going on with the rpmdb that's also slowing rpm down generally.

(I don't know if the rpmdb loader is asking SQLite to mmap the database in, but I fear it isn't, or it isn't making a big enough mapping to capture it.  That might also agree with the database size remarks.)

</details>

**Comment by ppisar** [comment]
My system has install less packages than yours. That could be the difference. So the problem appears not a repository size (primary.xml, libsolv files), but a number of installed packages (/usr/lib/sysimage/libdnf5/nevras.toml, /usr/lib/sysimage/rpm/rpmdb.sqlite).

If a speed of loading RPM database /usr/lib/sysimage/rpm/rpmdb.sqlite is a problem for you ("rpm -qa" took 8 seconds), I recommend asking for help RPM maintainers <https://github.com/rpm-software-management/rpm/issues>.

Though I wouldn't be surprised if the biggest difference between DNF4 and DNF5 consisted of parsing the TOML file (/usr/lib/sysimage/libdnf5/nevras.toml). At the end it's a plain text file.

**Comment by m-blaha** [comment]
There definitely is something terribly wrong - the larger package set is, the slower it runs. So for example `dnf5 info` prints results at pace approximately 1 package per second. How about dropping RFE and bumping the priority? I think this deserves some investigation.

**Comment by j-mracek** [comment]
I saw the performance of DNF info and this is unusable

`dnf info` requires 9s to finish, but `dnf5 info` requires ... I am sorry, it still running hopefully less then hour. 

I am proposing to add a high priority to this issue.

**Comment by kontura** [comment]
I made a PR with a workaround fix: https://github.com/rpm-software-management/dnf5/pull/1286

I would like to do a proper fix for dnf-5.2.0.0.
